New loader: Load PNG/TIF/JPG microscopy files as Nibabel objects

mariehbourget opened the pull request:
## Checklist

#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
This PR aims to load PNG microscopy files for training and load PNG/TIF/JPG files as Nibabel objects for inference.
Earlier in PR #734, those files were load as numpy arrays. However, we decided to change the loader for Nibabel objects to facilitate the integration of microscopy in the inference pipeline (reference thread in issue #306).

## TODO:
- [x] Load png/tif/jpg as Nibabel objects
- [x] Set "zooms" aka "pixdim" in Nibabel object based on PixelSize metadata from json sidecar files
- [x] Check if Nibabel objects need to be saved as NifTI files for inference pipeline --> answer: yes
- [x] If yes, save the NifTI files in an appropriate data structure --> Nifti files are written directly in the bids dataset alongside PNGs.

## Linked issues
Fixes #812

mariehbourget committed:
Remove apply_canonical function, revert to nib.as_closest_canonical

mariehbourget committed:
Remove get_voxel_size, revert to header.get_zooms

mariehbourget committed:
Remove get_shape, revert to header.get_data_shape

mariehbourget committed:
Remove get_data, revert to nibabel get_fdata

mariehbourget committed:
In read_file, convert png/tif/jpg to Nifti1Image object

mariehbourget committed:
Get pixel size from json metadata and set zooms in nifti header

mariehbourget committed:
Merge branch 'master' into mhb/812-loader-png-to-nifti

mariehbourget committed:
Merge branch 'master' into mhb/812-loader-png-to-nifti

jcohenadad commented on ivadomed/loader/loader.py:
```diff
@@ -187,8 +187,8 @@ class SegmentationPair(object):
         prepro_transforms (dict): Transforms to be applied before training.
-        input_handle (list): List of input nifty data as 'nibabel.nifti1.Nifti1Image' object or png/tif/jpg data as 'ndarray'
-        gt_handle (list): List of gt nifty data as 'nibabel.nifti1.Nifti1Image' object or png/tif/jpg data as 'ndarray'
+        input_handle (list): List of input nifty data as 'nibabel.nifti1.Nifti1Image' object
```
```suggestion
        input_handle (list): List of input NIfTI data as 'nibabel.nifti1.Nifti1Image' object
```

mariehbourget replied:
Fixed in 6cce642

jcohenadad commented on ivadomed/loader/loader.py:
```diff
@@ -187,8 +187,8 @@ class SegmentationPair(object):
-        input_handle (list): List of input nifty data as 'nibabel.nifti1.Nifti1Image' object or png/tif/jpg data as 'ndarray'
-        gt_handle (list): List of gt nifty data as 'nibabel.nifti1.Nifti1Image' object or png/tif/jpg data as 'ndarray'
+        input_handle (list): List of input nifty data as 'nibabel.nifti1.Nifti1Image' object
+        gt_handle (list): List of gt nifty data as 'nibabel.nifti1.Nifti1Image' object
```
```suggestion
        gt_handle (list): List of gt (ground truth) NIfTI data as 'nibabel.nifti1.Nifti1Image' object
```

mariehbourget replied:
Fixed in 6cce642

mariehbourget committed:
loader/utils.py: Correction to filtering of extensions in bids_dataframe
(was not filtering derivatives extensions before)

mariehbourget committed:
Update read_file in SegmentationPair to convert PNG to Nibabel object
and write Nifti file

mariehbourget committed:
configuration_file.rst: Update extensions definition and example

mariehbourget committed:
Move file extension check from SegmentationPair to get_file_extension in
loader/utils.py

mariehbourget committed:
Add update_filename_to_nifti function to loader/utils.py

mariehbourget committed:
test_loader.py: update test_load_dataset_2d_png

mariehbourget committed:
testing.py: Check extentions and update paths if not NifTI in get_gt

mariehbourget committed:
evaluation.py: Check extentions and update paths if not NifTI in
evaluate

mariehbourget committed:
inference.py: Check extentions and update paths if not NifTI in
pred_to_nib

mariehbourget committed:
Add TODO comment for inference in main.py

mariehbourget committed:
Add .tif and .tiff microscopy extensions to bids_dataframe

mariehbourget committed:
loader/utils.py: Correction to update_filename_to_nifti

mariehbourget committed:
Update config_new_loader.json to test the PR

mariehbourget committed:
Merge branch 'master' into mhb/812-loader-png-to-nifti

mariehbourget commented:
This PR is ready for review.

In `SegmentationPair --> read_file`:  png/tif/jpg files are converted as Nibabel object instead of numpy arrays, and written as NifTI files in the bids dataset.

In `testing.py`, `evaluate.py` and `inference.py`, a new function `update_filename_to_nifti` from `loader/utils.py` is called to update the filename list with the new NifTI path to allow inference with the current Nibabel/Nifti pipeline.

In order for the feature to work properly, I also included a correction to the filtering of the `bids_dataframe` based on `extensions`, and moved the verification of supported file extensions to a new `get_file_extension` function in `loader/utils.py` instead of directly in `SegmentationPair`.

To test the PR:
- Run a short training (CPU friendly) with `config_new_loader.json` and the [data_example_microscopy_sem](https://github.com/ivadomed/data_example_microscopy_sem) dataset.
- Modify the command in config to `test`and run the test.
- Both training and testing should run without errors.
- New NifTI files should be in the dataset alongside the PNGs.
- Note: the `test_fraction` in the config file is 0.2, there should be `preds_mask` for 2 files after the test but there will be only one because of another issue (#823) with inference, this is normal for now (out of scope for this PR), test_fraction of 0.1 would also returns an error for the same reason.
- Note 2: The command `segment` won't work because of missing metadata, also normal, I'll take care of it in issue #306.

Note to the reviewer, the diff from `loader.py` is kind of hard to read. Here is the big picture of the changes for this file:
1. The verification of supported file extension is moved to `loader/utils.py --> get_file_extension`
2. I removed 4 functions from PR #734 that were dealing with numpy arrays for PNG and reverted back to the previous behavior with Nibabel objects. Those are `apply_canonical`, `get_shape`, `get_voxel_size` and `get_data`.
2. I modified `read_file` to include the conversion of PNG to Nibabel with the PixelSize from the metadata (previously in `get_voxel_size`), and write the NifTI files.

joshuacwnewton commented on ivadomed/loader/utils.py:
```diff
@@ -654,18 +661,20 @@ def create_bids_dataframe(self):
                                | (df_next['path'].str.contains('derivatives')
                                & df_next['filename'].str.contains('|'.join(self.target_suffix)))]
 
+            # Update dataframe with files of chosen extensions if present
+            if self.extensions:
+                df_next = df_next[df_next['filename'].str.endswith(tuple(self.extensions))]
```
Nitpick: We can still use `df_next['extension']` here, right?

mariehbourget replied:
No we cannot use `['extension']` here unfortunately. This is because of the parsing from `pybids`.

When the files are "raw data" in a `sub-` folder, `pybids` parses all the entities + extensions of the filenames.
But when the files are in the `derivatives` folder, they are indexed but the filenames are not parsed and the `['extension']` cells are empty.
Example [here](https://github.com/ivadomed/data-testing/blob/master/microscopy_png/df_ref.csv): `seg-*` files don't have metadata in the dataframe.

I changed it to have the filtering by extensions on both on `sub-` and `derivatives`.

joshuacwnewton replied:
Thanks for the clarification! Sorry about that, that was a naive suggestion my part. :sweat_smile:

mariehbourget replied:
I removed the `if self.extensions` condition because now it is always true with `.nii` and `.nii.gz` by default.

mariehbourget committed:
Set self.extensions to nii and nii.gz by default in BidsDataframe when
no extensions are provided in config

mariehbourget committed:
Update docs/source/configuration_file.rst

Co-authored-by: Joshua Newton <joshuacwnewton@gmail.com>

mariehbourget commented on docs/source/configuration_file.rst:
```diff
@@ -287,7 +287,7 @@ will be randomly chosen.
         "title": "extensions",
         "description": "Used to specify a list of file extensions to be selected for
-            training/testing.",
+            training/testing. Extensions are required when different from `.nii` or `.nii.gz`",
```
Good catch! The behavior was ambiguous enough... it was indexing everything when no extensions were provided which was not the plan. I fixed it in 60d2fc6 and used your second suggestion in the doc, which was the behavior intended. Now, it indexes "nii" and "nii.gz" by default when the extensions key is missing or when the extensions list is empty.

mariehbourget commented on ivadomed/config/config_new_loader.json:
```diff
@@ -9,18 +9,18 @@
     "loader_parameters": {
-        "path_data": ["data_example_spinegeneric"],
+        "path_data": ["data_example_microscopy_sem"],
         "bids_config": "ivadomed/config/config_bids.json",
         "subject_selection": {"n": [], "metadata": [], "value": []},
-        "target_suffix": ["_seg-manual"],
-        "extensions": [],
+        "target_suffix": ["_seg-axon-manual", "_seg-myelin-manual"],
+        "extensions": [".png"],
```
You are correct. I change this config file with every PR for the loader/microscopy project for testing. When it will be completed, I will delete it and add an "official" microscopy config file.

mariehbourget commented on ivadomed/loader/utils.py:
```diff
@@ -835,3 +844,36 @@ def write_derivatives_dataset_description(self, path_data):
+    # Find the first match from the list of supported file extensions
+    extension = next((ext for ext in EXT_LST if filename.lower().endswith(ext)), None)
+
+    # TODO: remove "ome" from condition when implementing OMETIFF support (#739)
+    if (not extension) or ("ome" in extension):
+        raise RuntimeError("The input file type of '{}' is not supported".format(filename))
```
This is a good question with a quite long answer :)

1. I don't use `pathlib.Path.suffixes` because we have to account for files with two part extensions (ex: `nii.gz`), for files with similar extensions (we don't want to mistake a `tiff` with an `ome.tiff`) and for files that could include another dot in the filenames (could happen when calling directly `segment_volume`, ex: with `sct_deepseg`). This is why we use an ordered list of supported extensions and a `str.endswith`.

2. I agree that the validation should not be done everytime and I suggest to move the validation (`if (not extension) or ("ome" in extension):`) in `SegmentationPair --> read_file`. This way, the validation would be done only once when the file is first loaded.

3. I cannot do this in the `BidsDataframe` because some entry points don't use it. Basically, here are the different entry points:

- command train/test:  `BidsDataframe --> BidsDataset --> MRI2DSegmentationDataset --> SegmentationPair`
- command segment: `BidsDataframe  --> segment_volume --> MRI2DSegmentationDataset --> SegmentationPair`
- sct_deepseg (and soon ADS): --> `segment_volume --> MRI2DSegmentationDataset --> SegmentationPair`

Same 3 options with 3D loader and HDF5 (adaptative) loader.
The one thing that they have in common is that they all go through `SegmentationPair --> read_file`. To avoid any code duplication, I think this is the best place to do the check.

Let me know what you think!

joshuacwnewton replied:
> * sct_deepseg (and soon ADS): --> `segment_volume --> MRI2DSegmentationDataset --> SegmentationPair`

Aha, this is the part that I was missing. I had been assuming that everything went through `BidsDataframe`. Thank you giving such an in-depth answer! :tada:

mariehbourget replied:
I moved the validation from `get_file_extension` to `read_file` after fetching the extension.

joshuacwnewton commented on ivadomed/loader/utils.py:
```diff
@@ -29,6 +29,13 @@
 TRANSFORM_PARAMS = ['elastic', 'rotation', 'scale', 'offset', 'crop_params', 'reverse',
                     'translation', 'gaussian_noise']
 
+# Ordered list of supported file extensions
+# TODO: Implement support of the following OMETIFF formats (#739):
+# [".ome.tif", ".ome.tiff", ".ome.tf2", ".ome.tf8", ".ome.btf"]
+# They are included in the list to avoid a ".ome.tif" or ".ome.tiff" following the ".tif" or ".tiff" pipeline
+EXT_LST = [".nii", ".nii.gz", ".ome.tif", ".ome.tiff", ".ome.tf2", ".ome.tf8", ".ome.btf", ".tif",
+           ".tiff", ".png", ".jpg", ".jpeg"]
```
Would it be valuable to expose this list to the user somehow? Either through the documentation ("`extensions` supports the following..."), or in the error message, or both?

mariehbourget commented on ivadomed/testing.py:
```diff
@@ -410,6 +410,9 @@ def get_gt(filenames):
     Returns:
         ndarray: 4D numpy array.
     """
+    # Check filenames extentions and update paths if not NifTI
+    filenames = [imed_loader_utils.update_filename_to_nifti(fname) for fname in filenames]
+
```
I tried it earlier this week and it's a whole other can of worms. 
So it is possible but it also requires several changes to the code that were not better in my opinion.

1. First step would be returning and updating the filenames with the function `read_file` [here](https://github.com/ivadomed/ivadomed/blob/9e3dfdacf2c38bf452a51991ed7385ae4c897043/ivadomed/loader/loader.py#L209), [here](https://github.com/ivadomed/ivadomed/blob/9e3dfdacf2c38bf452a51991ed7385ae4c897043/ivadomed/loader/loader.py#L224) and [here](https://github.com/ivadomed/ivadomed/blob/9e3dfdacf2c38bf452a51991ed7385ae4c897043/ivadomed/loader/loader.py#L226). Then updating the `self.metadata` [here](https://github.com/ivadomed/ivadomed/blob/9e3dfdacf2c38bf452a51991ed7385ae4c897043/ivadomed/loader/loader.py#L251) with the new filenames. That would work for `train` and `test` without any changes in `testing.py` and `inference.py`, but would require changes to `evalution.py` (point 2).

2. For `evaluation.py`, the solution would remain the same as the one I implemented because the filenames are taken [here](https://github.com/ivadomed/ivadomed/blob/9e3dfdacf2c38bf452a51991ed7385ae4c897043/ivadomed/evaluation.py#L53) from the `bids_dataframe` which does not have the updated filenames. We do not want to update the `bids_dataframe` to keep a  reference to the original files and metadata.

3.  For the `segment` command and through the `segment_volume` (sct_deepseg) entry point, we would need changes to `inference.py`. We would have to fetch the new filenames from the metadata in the DataLoader batch (ex: `batch['input_metadata'][0][0]['input_filenames']`) somewhere around [here](https://github.com/ivadomed/ivadomed/blob/9e3dfdacf2c38bf452a51991ed7385ae4c897043/ivadomed/inference.py#L497) to pass to `pred_to_nib`, instead of using the filename list from the parameters.

So my conclusion after all of this was that updating the filenames was not worth it, it seemed to me that this solution was a lot harder to implement, understand and maintain.  Moreover, there would be discrepancies between the paths, filenames and extensions in the metadata  unless we update them all.
Again, let me know what you think, thanks!

(and now that I look at it again, I should probably use my `update_filename_to_nifti` function instead of the `filename.split('.')[0] + ".nii.gz"`).

joshuacwnewton replied:
Thanks for being so thorough about this! I'm convinced. :slightly_smiling_face:

mariehbourget replied:
> (and now that I look at it again, I should probably use my update_filename_to_nifti function instead of the filename.split('.')[0] + ".nii.gz").

Fixed

joshuacwnewton reviewed (commented):
Thanks for being so thorough in answering my questons about the bigger structure/organization of the PR. :smile: 

I only have a few smaller comments related to documentation/clarity now.

mariehbourget committed:
Move file extension validation to read_file in SegmentationPair

mariehbourget committed:
Use update_filename_to_nifti to update path in read_file

mariehbourget committed:
Remove if self.extensions condition in filtering of bids_dataframe
(always true)

mariehbourget committed:
Correction to PixelSize assignment and comments

mariehbourget committed:
Add convert_file_to_nifti method to SegmentationPair

mariehbourget committed:
Update update_filename_to_nifti docstings

Co-authored-by: Joshua Newton <joshuacwnewton@gmail.com>

mariehbourget committed:
Update read_file docstrings

mariehbourget committed:
Add list of supported file extensions in error message in read_file

mariehbourget committed:
Fix resize of PixelSize

mariehbourget commented on ivadomed/loader/utils.py:
```diff
@@ -29,6 +29,13 @@
 TRANSFORM_PARAMS = ['elastic', 'rotation', 'scale', 'offset', 'crop_params', 'reverse',
                     'translation', 'gaussian_noise']
 
+# Ordered list of supported file extensions
+# TODO: Implement support of the following OMETIFF formats (#739):
+# [".ome.tif", ".ome.tiff", ".ome.tf2", ".ome.tf8", ".ome.btf"]
+# They are included in the list to avoid a ".ome.tif" or ".ome.tiff" following the ".tif" or ".tiff" pipeline
+EXT_LST = [".nii", ".nii.gz", ".ome.tif", ".ome.tiff", ".ome.tf2", ".ome.tf8", ".ome.btf", ".tif",
+           ".tiff", ".png", ".jpg", ".jpeg"]
```
I added the list to the error message in 7e26faf.
I agree that we should also update the documentation but I think it would be best to do it in "data" [here](https://ivadomed.org/en/latest/data.html). I would prefer to wait until the microscopy pipeline is complete before doing that. I also plan on adding a tutorial for microscopy.

mariehbourget commented on ivadomed/loader/loader.py:
```diff
@@ -440,124 +426,65 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
-            return data.get_fdata(cache_mode, dtype=np.float32)
+            # Set "pixdim" (zooms) in Nifti1Image object header
+            # Note: pixdim[1,2,3] must be non-zero in Nifti objects even if there is only one slice.
+            # Above, when ps_in_um[2] (pixdim[3]) is 0, we assign the same PixelSize as ps_in_um[0] (pixdim[1])
```
The comment is accurate but the implementation was badly executed. I fixed the implementation and moved the comment at the appropriate place in fa5c2a9 and 8fb9c56.

mariehbourget commented on ivadomed/loader/utils.py:
```diff
@@ -835,3 +844,36 @@ def write_derivatives_dataset_description(self, path_data):
+
+
+def update_filename_to_nifti(filename):
+    """ Update filename extension to 'nii.gz' if not NifTI file.
```
Thank you for the suggestion, it is much clearer now. Updated in b83a6f3.

mariehbourget commented on ivadomed/loader/loader.py:
```diff
@@ -440,124 +426,65 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
 
         else:
             # For '.png', '.tif', '.tiff', '.jpg' and 'jpeg' extentions
-            # Voxel size is extracted from PixelSize metadata (from BIDS JSON sidecar)
-            # PixelSize definition in example dataset is a scalar in micrometers (BIDS BEP031 v 0.0.2)
-            # PixelSize definition may change for 2D [X, Y] and 3D [X, Y, Z] arrays in micrometers (BIDS BEP031 v 0.0.3)
-            # This method supports both behaviors.
-            # TODO: Update behavior to follow BEP microscopy development (#301)
```
The comments were still relevant, my mistake for removing them. I took your suggestion and implemented the `convert_file_to_nifti` method with appropriate docstrings in b8c435c.

mariehbourget commented:
Thank you @joshuacwnewton for the review and great feedback!
I made the corrections according to your suggestions.

mariehbourget commented:
@joshuacwnewton, are the changes OK with you? If yes, can you approve the PR? Thanks! :)

joshuacwnewton reviewed (approved):
LGTM! (Thanks for the reminder ping. :smile:)
